_posts/2016-04-19-MIR.md

```diff
@@ -0,0 +1,725 @@
+    *incremental*, so that when you *re*-compile code, the compiler recomputes only
+    what it has to. MIR has been designed from the start with this use-case in mind,
+    so it's much easier for us to save and reload, even if other parts of the program
+    have changed in the meantime.
+
+    MIR also provides a foundation for more efficient data structures and removal
+    of redundant work in the compiler, both of which should speed up compilation
+    across the board.
+
+-   **Faster execution time**. You may have noticed that in the new compiler
+    pipeline, *optimization* appears twice. That's no accident: previously, the
+    compiler relied solely on LLVM to perform optimizations, but with MIR, we can
+    do some *Rust-specific* optimizations before ever hitting LLVM -- or, for that
+    matter, before monomorphizing code. Rust's rich type system should provide
+    fertile ground for going beyond LLVM's optimizations.
+
+    In addition, MIR will uncork some longstanding performance improvements
+    to the code Rust generates, like ["non-zeroing" drop][rfc320].
+
+-   **More precise type checking**. Today's Rust compiler imposes some
+    artificial restrictions on *borrowing*, restrictions which largely
+    stem from the way the compiler currently represents programs. MIR
+    will enable [much more flexible borrowing][811], which will in turn
+    improve Rust's ergonomics and learning curve.
+
+Beyond these banner user-facing improvements, **MIR also has substantial
+engineering benefits for the compiler**:
+
+- **Eliminating redundancy.** Currently, because we write all of our passes in
+  terms of the full Rust language, there is quite a lot of duplication. For
+  example, both the safety analyses and the backend which produces LLVM IR must
+  agree about how to translate drops, or the precise order in which `match`
+  expression arms will be tested and executed (which can get quite
+  complex). With MIR, all of that logic is centralized in MIR construction, and
+  the later passes can just rely on that.
+
+- **Raising ambitions.** In addition to being more [DRY][dry], working
+  with MIR is just plain *easier*, because it contains a much more
+  primitive set of operations than ordinary Rust. This simplification
+  enables us to do a lot of things that were forbodingly complex
+  before. We'll look at one such case in those post -- non-zeroing
+  drop -- but as we'll see at the end, there are already many others
+  in the pipeline.
+
+Needless to say, we're excited, and the Rust community has stepped up
+in a big way to make MIR a reality. The compiler can bootstrap and run
+its test suite using MIR, and these tests have to pass on every new
+commit. Once we're able to run [Crater][crater] with MIR enabled and
+see no regressions across the entire [crates.io](http://crates.io/)
+ecosystem, we'll turn it on by default (or, you'll forgive a terrible
+(wonderful) pun, [launch MIR into orbit][orbit]).
```

Comment:
I would cut this para or replace it with something much shorter - "MIR is in the compiler, will be turned on soon-ish"